ppi.py

```diff
@@ -0,0 +1,89 @@
+from binaryninja import BinaryView, MediumLevelILCall, MediumLevelILTailcall, MediumLevelILLoadStruct, PointerType, \
+    MediumLevelILConstPtr, BackgroundTask
+from .protocols import lookup_protocol_guid, define_protocol_types
+from .system_table import propagate_system_table_pointers
+
+
+def define_descriptor(bv: BinaryView, task):
+    descriptor_types = ["EFI_PEI_NOTIFY_DESCRIPTOR", "EFI_PEI_PPI_DESCRIPTOR", "EFI_PEI_DESCRIPTOR", ]
+    for descriptor_type in descriptor_types:
+        refs = list(bv.get_code_refs_for_type(descriptor_type))
+        for ref in refs:
+            if task.cancelled:
+                return False
+            instr = ref.mlil
+            if isinstance(instr, MediumLevelILCall) or isinstance(instr, MediumLevelILTailcall):
+                if isinstance(instr.dest, MediumLevelILLoadStruct):
+                    # likely a call to services
+                    if isinstance(instr.dest.expr_type, PointerType):
+                        function_type = instr.dest.expr_type.target
+                        params = function_type.parameters
+                        target_param = None
+                        for param in params:
+                            type_name = str(param.type)
+                            if "DESCRIPTOR" in type_name:
+                                target_param = params.index(param)
+                                break
+                        if target_param is None:
+                            continue
+
+                        param = instr.params[target_param]
+                        if not isinstance(param, MediumLevelILConstPtr):
+                            continue
+
+                        var_addr = param.constant
+                        sym = bv.get_symbol_at(var_addr)
+
+                        if sym is not None:
+                            var_name = sym.name
+                        else:
+                            var_name = None
+
+                        if type_name[-1] == '*':
```

Comment:
We should do this more intelligently in case the type is something like `EFI_PEI_NOTIFY_DESCRIPTOR**` if the intentions are to just get the type name of `EFI_PEI_NOTIFY_DESCRIPTOR`. Likely a way to do this using the BN API. Or at the very least you could do somethling like:

```python
if type_name.endswith("*"):
    type_name = type_name.rstrip("*")
```